### ISSUE TITLE
Linking error with @tailrec method inside a trait with a self-type, on Scala 2.12+

### ISSUE THREAD

**Issue by oyvindberg** [scala-js/scala-js#3267]
We stumbled upon a case which the linker doesn't seem to handle while porting http4s ( https://github.com/http4s/http4s/pull/1412 ). 

I made an as minimal repro as I could manage in https://github.com/oyvindberg/scalajs-linker-repro/blob/master/src/main/scala/http4s/Tester.scala . 

As you can see there are several features at work here (macros (`rule`), self types, inheritance), and there are two acceptable workarounds to make it run. It's not a blocker for the port, but it certainly cost me enough headache to figure out that I figured I should report it properly.

Error message is
```
sbt:repro> run
[info] Compiling 1 Scala source to /home/olvind/pr/scalajs-linker-repro/target/scala-2.12/classes ...
[info] Done compiling.
[info] Fast optimizing /home/olvind/pr/scalajs-linker-repro/target/scala-2.12/repro-fastopt.js
[error] Referring to non-existent method http4s.Tester$$anon$2.b()org.parboiled2.Rule
[error]   called from http4s.Tester$Helpers.rec$2(scala.Int,scala.Long,scala.collection.immutable.VectorBuilder,scala.Int,scala.Int)scala.Boolean
[error]   called from http4s.Tester$$anon$1.rec$2(scala.Int,scala.Long,scala.collection.immutable.VectorBuilder,scala.Int,scala.Int)scala.Boolean
[error]   called from http4s.Tester$Helpers.a()org.parboiled2.Rule
[error]   called from http4s.Tester$$anon$1.a()org.parboiled2.Rule
[error]   called from http4s.Tester$.$$anonfun$one$1(http4s.Tester$$anon$1)org.parboiled2.Rule
[error]   called from http4s.Tester$.one(java.lang.String)scala.collection.Seq
[error]   called from http4s.Tester$.main([java.lang.String)scala.Unit
[error]   called from core module module initializers
[error] involving instantiated classes:
[error]   http4s.Tester$$anon$1
[error]   http4s.Tester$
[error] Referring to non-existent method http4s.Tester$$anon$2.b()org.parboiled2.Rule
[error]   called from http4s.Tester$Helpers.rec$1(scala.Int,scala.Long,scala.collection.immutable.VectorBuilder,scala.Int,scala.Int)scala.Boolean
[error]   called from http4s.Tester$$anon$1.rec$1(scala.Int,scala.Long,scala.collection.immutable.VectorBuilder,scala.Int,scala.Int)scala.Boolean
[error]   called from http4s.Tester$Helpers.wrapped$2()scala.Boolean
[error]   called from http4s.Tester$$anon$1.wrapped$2()scala.Boolean
[error]   called from http4s.Tester$Helpers.a()org.parboiled2.Rule
[error]   called from http4s.Tester$$anon$1.a()org.parboiled2.Rule
[error]   called from http4s.Tester$.$$anonfun$one$1(http4s.Tester$$anon$1)org.parboiled2.Rule
[error]   called from http4s.Tester$.one(java.lang.String)scala.collection.Seq
[error]   called from http4s.Tester$.main([java.lang.String)scala.Unit
[error]   called from core module module initializers
[error] involving instantiated classes:
[error]   http4s.Tester$$anon$1
[error]   http4s.Tester$
[error] org.scalajs.core.tools.linker.LinkingException: There were linking errors
[error]         at org.scalajs.core.tools.linker.frontend.BaseLinker.linkInternal(BaseLinker.scala:160)
[error]         at org.scalajs.core.tools.linker.frontend.BaseLinker.linkInternal(BaseLinker.scala:108)
[error]         at org.scalajs.core.tools.linker.frontend.LinkerFrontend.$anonfun$link$3(LinkerFrontend.scala:63)
[error]         at org.scalajs.core.tools.logging.Logger.time(Logger.scala:28)
[error]         at org.scalajs.core.tools.logging.Logger.time$(Logger.scala:26)
[error]         at org.scalajs.sbtplugin.Loggers$SbtLoggerWrapper.time(Loggers.scala:7)
[error]         at org.scalajs.core.tools.linker.frontend.LinkerFrontend.link(LinkerFrontend.scala:62)
[error]         at org.scalajs.core.tools.linker.Linker.$anonfun$link$1(Linker.scala:52)
[error]         at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error]         at org.scalajs.core.tools.linker.Linker.guard(Linker.scala:69)
[error]         at org.scalajs.core.tools.linker.Linker.link(Linker.scala:50)
[error]         at org.scalajs.core.tools.linker.ClearableLinker.$anonfun$link$1(ClearableLinker.scala:52)
[error]         at org.scalajs.core.tools.linker.ClearableLinker.$anonfun$link$1$adapted(ClearableLinker.scala:52)
[error]         at org.scalajs.core.tools.linker.ClearableLinker.linkerOp(ClearableLinker.scala:63)
[error]         at org.scalajs.core.tools.linker.ClearableLinker.link(ClearableLinker.scala:52)
[error]         at org.scalajs.sbtplugin.ScalaJSPluginInternal$.$anonfun$scalaJSStageSettings$11(ScalaJSPluginInternal.scala:310)
[error]         at sbt.util.FileFunction$.$anonfun$cached$1(FileFunction.scala:73)
[error]         at sbt.util.FileFunction$.$anonfun$cached$4(FileFunction.scala:147)
[error]         at sbt.util.Difference.apply(Tracked.scala:313)
[error]         at sbt.util.Difference.apply(Tracked.scala:293)
[error]         at sbt.util.FileFunction$.$anonfun$cached$3(FileFunction.scala:143)
[error]         at sbt.util.Difference.apply(Tracked.scala:313)
[error]         at sbt.util.Difference.apply(Tracked.scala:288)
[error]         at sbt.util.FileFunction$.$anonfun$cached$2(FileFunction.scala:142)
[error]         at org.scalajs.sbtplugin.ScalaJSPluginInternal$.$anonfun$scalaJSStageSettings$10(ScalaJSPluginInternal.scala:315)
[error]         at sbt.std.Transform$$anon$3.$anonfun$apply$2(System.scala:46)
[error]         at sbt.std.Transform$$anon$4.work(System.scala:66)
[error]         at sbt.Execute.$anonfun$submit$2(Execute.scala:262)
[error]         at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error]         at sbt.Execute.work(Execute.scala:271)
[error]         at sbt.Execute.$anonfun$submit$1(Execute.scala:262)
[error]         at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:174)
[error]         at sbt.CompletionService$$anon$2.call(CompletionService.scala:36)
[error]         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error]         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error]         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error]         at java.lang.Thread.run(Thread.java:748)
[error] (Compile / fastOptJS) org.scalajs.core.tools.linker.LinkingException: There were linking errors
```

**Comment by sjrd** [comment]
Minimization that does not involve any library nor macro:
```scala
package helloworld

import scala.annotation.tailrec

class Parser

trait Helpers { this: Parser =>
  @tailrec
  final def rec(i: Int): Int = {
    if (i == 0) b()
    else rec(i - 1)
  }

  def b(): Int = 42
}

class A extends Parser {
  def foo(): Int = 0
}

class B extends Parser with Helpers

object HelloWorld {
  def main(args: Array[String]): Unit = {
    println(new A().foo())
    println(new B().rec(3))
  }
}
```
The IR of `Helpers.rec()` is faulty:
```scala
interface Lhelloworld_Helpers {
  def rec__I__I(var i: int): int = {
    var _$this: Lhelloworld_Parser = this.asInstanceOf[Lhelloworld_Parser];
    x[int]: {
      _rec: while (true) {
        return(x) if ((i == 0)) {
          _$this.b__I()
        } else {
          val temp$_$this: Lhelloworld_Parser = _$this.asInstanceOf[Lhelloworld_Parser];
          val temp$i: int = (i -[int] 1);
          _$this = temp$_$this;
          i = temp$i;
          continue _rec
        }
      }
    }
  }
  def b__I(): int = {
    42
  }
  def $$init$__V() {
    /*<skip>*/
  }
}
```
It calls `_$this.b__i()`, but `_$this` is of type `Parser`, which does not have a method `b__I`.

`-Xprint:mixin` tells a different story:
```scala
  abstract trait Helpers extends Object { `this`: helloworld.Parser => 
    @scala.annotation.tailrec final def rec(i: Int): Int = {
      <synthetic> val _$this: helloworld.Parser = Helpers.this.$asInstanceOf[helloworld.Parser]();
      _rec(_$this: helloworld.Parser, i: Int){
        (if (i.==(0))
          Helpers.this.b()
        else
          _rec(Helpers.this.$asInstanceOf[helloworld.Parser](), (i.-(1): Int)): Int)
      }
    };
    def b(): Int = 42;
    def /*Helpers*/$init$(): Unit = {
      ()
    }
  };
```
Here we call `Helpers.this.b()`, which is well-typed. Our codegen replaces that `Helpers.this` by `_$this` because, well, the tailrec treatment leaves abominable traces that we have to deal with.

This is clearly related to #3058. It is now pretty clear that the fix in #3059 was wrong. We should not believe scalac regarding the type of `_$this`: it is considered as a `Helpers` everywhere but at definition site, not as a `Parser`. This can be seen in the `_rec()` call, and also if we call a method of `Parser`: scalac emits an explicit `_$this.asInstanceOf[Parser]` anyway.

scalac, you are really doing nasty stuff sometimes.

**Comment by sjrd** [comment]
Fixed in a0802f444de87f021f0828997caa0786a8abc289.

Thank you @oyvindberg for the extremely helpful bug report and minimization!

**Comment by oyvindberg** [comment]
Thanks to you @sjrd, I'm super impressed that you managed to fix such a complex looking bug so quickly :)
